sympy/sets/conditionset.py

```diff
@@ -246,6 +246,33 @@ def _eval_subs(self, old, new):
             return ConditionSet(new, Contains(new, base), base)
         return self.func(self.sym, cond, base)
 
+    def _eval_rewrite_as_Intersection(self, *args, **kwargs):
+        """
+        Try to convert the ConditionSet to an Intersection set.
+
+        Examples
+        ========
+
+        >>> from sympy import  S, ConditionSet, Intersection
+        >>> from sympy.abc import x
+
+        >>> ConditionSet(x, x<1, S.Reals).rewrite(Intersection)
+        Interval.open(-oo, 1)
+
+        >>> ConditionSet(x, x**2 > x, S.Integers).rewrite(Intersection)
+        Union(Range(-oo, 0, 1), Range(2, oo, 1))
+
+        Notes
+        =====
+        Currently this is only implemented for ConditionSets which have a single variable and where the base set
+        is a subset of `Reals`.
+        """
+        if not isinstance(self.sym, Tuple) and self.base_set.is_subset(S.Reals):
+            try:
+                return self.condition.as_set().intersect(self.base_set)
+            except NotImplementedError:  # may be raised in solve_univariate_inequality
```

Comment:
The above should not raise. Which part of it raises? That should be fixed there rather than here.

Reply:
In the tests, the following call is made `Eq(_n*cos(_n) - 3*sin(_n), 0).as_set()` .
Stacktrace:
```
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
~/workspace/sympy/sympy/solvers/inequalities.py in solve_univariate_inequality(expr, gen, relational, domain, continuous)
    523                 # or it might give None...
--> 524                 solns = solvify(e, gen, domain)
    525                 if solns is None:

~/workspace/sympy/sympy/solvers/solveset.py in solvify(f, symbol, domain)
   2132     elif isinstance(solution_set, ConditionSet):
-> 2133         raise NotImplementedError('solveset is unable to solve this equation.')
   2134 

NotImplementedError: solveset is unable to solve this equation.

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
<ipython-input-5-c50d9d0e5bb9> in <module>
----> 1 Eq(_n*cos(_n) - 3*sin(_n), 0).as_set()

~/workspace/sympy/sympy/logic/boolalg.py in as_set(self)
    157                         with periodic solutions
    158                         '''))
--> 159             return self.subs(reps)._eval_as_set()
    160         else:
    161             raise NotImplementedError("Sorry, as_set has not yet been"

~/workspace/sympy/sympy/core/relational.py in _eval_as_set(self)
    385         x = syms.pop()
    386         import ipdb; ipdb.set_trace()
--> 387         return solve_univariate_inequality(self, x, relational=False)
    388 
    389     @property

~/workspace/sympy/sympy/solvers/inequalities.py in solve_univariate_inequality(expr, gen, relational, domain, continuous)
    532                     The inequality, %s, cannot be solved using
    533                     solve_univariate_inequality.
--> 534                     ''' % expr.subs(gen, Symbol('x'))))
    535 
    536             expanded_e = expand_mul(e)

NotImplementedError: 
The inequality, Eq(x*cos(x) - 3*sin(x), 0), cannot be solved using
solve_univariate_inequality.
```
I don't have time to dig further atm

Reply:
The problem is
```julia
In [10]: Eq(n*cos(n) - 3*sin(n), 0).as_set()                                                                                                                  
---------------------------------------------------------------------------
NotImplementedError
```
Here `solveset` raises `NotImplementedError` but probably a `ConditionSet` should be returned by `solveset` instead. The obvious result of `as_set()` here is
```julia
In [11]: ConditionSet(n, Eq(n*cos(n) - 3*sin(n), 0), Reals)                                                                                                   
Out[11]: {n | n ∊ ℝ ∧ n⋅cos(n) - 3⋅sin(n) = 0}
```

Reply:
That doesn't need to be fixed in this PR but I don't want the code here to catch the exception if the exception only comes from a bug. That just hides the bug.

You could add an XFAIL test instead and open an issue.